File: nodejs/data_collector_nodejs/master.js
// require("dotenv").config();
const express = require("express");
const morgan = require("morgan");
const { Sequelize, DataTypes, Model } = require("sequelize");

const DB_PATH = "./db/regs.db";
const TASK_LIMIT = 50000;

// const DB_PATH = "./db/test.db";
// const TASK_LIMIT = 10;

const sequelize = new Sequelize("database", "username", "password", {
  dialect: "sqlite",
  storage: DB_PATH,
  logging: false,
});

class Regs extends Model {}

Regs.init(
  {
    reg: {
      type: DataTypes.INTEGER,
      allowNull: false,
      unique: true,
      primaryKey: true,
    },
    data: {
      type: DataTypes.JSON,
      defaultValue: null,
    },
    collected: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
    },
  },
  {
    sequelize,
    modelName: "Regs",
    tableName: "Regs",
    timestamps: false,
  }
);

const app = express();

class Tasks {
  constructor(limit) {
    this.all = [];
    this.count = 0;
    this.limit = limit;
  }
  async begin() {
    this.all = await Regs.findAll({
      where: {
        collected: false,
      },
      limit: this.limit,
    });
  }

  async next() {
    const last = this.all[this.count];
    const index = this.count;
    this.count++;
    if (this.count > this.limit) {
      return null;
    }
    last.set("collected", true);
    await last.save();
    return [last.reg, index];
  }

  toJSON() {
    return this.all;
  }
}

const tasks = new Tasks(TASK_LIMIT);

app.use([
  morgan("dev"),
  express.urlencoded({ extends: true, limit: "500mb" }),
  express.json({ limit: "500mb" }),
]);

app.get("/", async (req, res) => {
  try {
    res.status(200).json(tasks.toJSON());
  } catch (e) {
    console.log(e);
    res.status(400).send("WTF");
  }
});

app.get("/ok", (req, res) => {
  res.status(200).json({ state: true, msg: "ALL OK" });
});

app.get("/c/n", async (req, res) => {
  const next = await tasks.next();
  if (next) {
    return res.status(200).send(`${next[0]}, ${next[1]}`);
  }
  res.status(400).json(null);
});

app.post("/v/:reg/:index", async (req, res) => {
  const { reg, index } = req.params;
  const { data } = req.body;

  if (!req) {
    return res.status(400).json(null);
  }

  const reg_data = await Regs.findByPk(reg);
  if (!reg_data) {
    return res.status(400).send(`${reg}`);
  }
  tasks.all[Number(index)].data = data;
  reg_data.set("data", data);
  await reg_data.save();
  res.status(200).json(reg_data);
});

const PORT = 8200;

sequelize
  .sync()
  .then(async () => {
    console.log("DB Connected");

    await tasks.begin();
    console.log(tasks.all.length, "Task loaded");
    app.listen(PORT, (err) => {
      if (err) {
        console.log(err);
        console.log("Server failed.");
      } else {
        console.log(`http://localhost:${PORT}`);
      }
    });
  })
  .catch((e) => {
    console.log(e);
    console.log("Faild to connecte DB");
  });

File: nodejs/data_collector/master.py
from flask import Flask, request,  jsonify
import sqlite3
import json

app = Flask(__name__)
TASK_LIMIT = 5
DB_PATH = "./db/test.db"


class Tasks:
    def __init__(self, task_limit, db_path) -> None:
        self.__count = 0
        self.all = []
        self.db_path = db_path
        self.__task_limit = task_limit
        self.__conn = sqlite3.connect(self.db_path)
        self.__db = self.__conn.cursor()
        self.all = self.__db.execute(
            f"SELECT reg,collected FROM Regs WHERE collected=0 LIMIT {self.__task_limit}").fetchall()
        self.__conn.close()
        # print(self.all)

    def next(self):
        last = self.__count

        if self.__count >= self.__task_limit:
            return None

        self.__count += 1
        return f'{self.all[last][0]},{last}'

    def this_is_valid(self, reg: int, index: int, mdata):
        data = self.all[index]
        self.all[index] = (data[0], 1)
        conn = sqlite3.connect(self.db_path)
        cur = conn.cursor()
        cur.execute(
            '''UPDATE Regs SET collected = ?, data = ? WHERE reg = ?''', (True, json.dumps(mdata), reg))
        conn.commit()
        conn.close()

        return "OK"

    def json(self):
        def c_json(data):
            return {
                "reg": data[0],
                "collected": bool(data[1]),
            }
        return list(map(c_json, self.all))


tasks = Tasks(task_limit=TASK_LIMIT, db_path=DB_PATH)


@app.route("/c", methods=['GET'])
def get_count_curr():
    return jsonify(tasks.json()), 200


@app.route("/c/n", methods=['GET'])
def get_count_next():
    data = tasks.next()
    if not data:
        return str(None), 400
    return str(data), 200


@app.route("/v/<int:reg>/<int:ind>", methods=["POST"])
def post_verify(reg, ind):
    data = request.form["data"]
    tasks.this_is_valid(reg, ind, data)
    return "OK",200


@app.route("/ok", methods=["GET"])
def server_ok():
    return jsonify({"msg": "OK", "state": True}), 200


if __name__ == '__main__':
    app.run(debug=False)
